src/Validation/Validation.php

```diff
@@ -353,7 +353,7 @@ public static function compareFields($check, string $field, string $operator, ar
      */
     public static function containsNonAlphaNumeric($check, int $count = 1): bool
     {
-        if (!is_scalar($check)) {
+        if (!is_string($check)) {
```

Comment:
This should not be changed. (int)-1 will be now return false, for example. Also, objects which implement __toString() will be no longer pass (It would be very rare case though). Instead, cast to string before passing to preg_match().
The same can be said to other is_string() check added in this change.

Reply:
POST data can only be string or array, so it really worth it worry about other types?
Also trying to type cast an array to string generates notice `Notice: Array to string conversion` https://3v4l.org/iTHHB.

So `is_string()` check is the right thing to do IMO.

Reply:
For others method like `luhn()`, `latitude()` etc. where numeric value is acceptable there `is_scalar()` check followed by string typecasting is used.

Reply:
Cant it also come from json decoded data and from api data?

Reply:
No. I think you may be surprised, but cake applications allow users to POST float/integer/boolean by default.

Reply:
Sorry, @dereuromark has been already mentioned it earlier than me 😛

Reply:
I overlooked the fact that it could be json decoded data.

That said why would anyone intentionally want to use "containsNonAlphaNumeric" for non string data?

Similarly other methods like max/minLength() are only meant for string and should reject non string data.

Reply:
> That said why would anyone intentionally want to use containsNonAlphaNumeric() for non string data?

Now it works so. If we want to change the behavior, we would need to leave it our migration guide.

> Similarly other methods like max/minLength() are only meant for string and should reject non string data.

Those can be used to restrict number of digits for integer or float values. So I don't think it is a good idea to change the behaviors.

Reply:
> Now it works so. If we want to change the behavior, we would need to write it in our migration guide.

The method checks if string contains one or more non-alphanumeric characters. So integers are going to fail validation even after conversion to string and I don't think any sane individual would expect this method to convert float value like `1.2` to string and then let it pass. So there's no change in behavior which needs to be documented.

> Those can be used to restrict number of digits for integer or float values. So I don't think it is a good idea to change the behaviors.

I'll update those to allow strings or numeric values.

Reply:
There's actually existing test case for `lengthBetween()` which fails for integer input.

https://github.com/cakephp/cakephp/blob/abbb678b23de16d0e4d15367804d7ae741510842/tests/TestCase/Validation/ValidationTest.php#L171

Hence `min/maxLength()` should also not accept integers.

Reply:
The is_string() checks in lengthBetween() was added in [a security patch](https://github.com/cakephp/cakephp/commit/c4c27018bccde76ce3955b36430febd1786e62c5) in the past. In the first version of the patch, it used is_string(). But since it would break existing apps, I said:
> I think that type checks in the cc() and luhn() would be better `!is_scalar` than `!is_string` like a naturalNumber(). 

And those are changed to is_string(). But I noticed that I overlooked lengthBetween(), and sent another mail:
> In `lengthBetween`, it also might be better `is_scalar` than `is_string`.
For example, "Please enter valid a number length between 3 and 5 digits".
Sorry, I overlooked the change in your previous patch. The other changes seem good to me.

But it seemed to be not applied to the final version of the patch. So the is_string() check in lengthBetween()  was not intended. 

So please do not change the current behavior of minLength() and maxLength(). I would rather want to change is_string() to is_scalar() in the lengthBetween() even now.

Reply:
In over 2 years since that patch no one has had a problem with that method only accepting strings :)

Anyway, for all length changes methods I have changed to use `is_scalar()`.

Reply:
I think application/json input is not used much. anyway, thank you!

Reply:
It can also be user added data via setters on the entity which then get validated.
However you generate the entities, they must not fail because of this.

Reply:
It is IMO :)

Reply:
Dont you think this is a dangerous behavior change?

Reply:
> Is this good to merge now?

No, I found a mistake.

Reply:
> Dont you think this is a dangerous behavior change?

@dereuromark I don't. I already explained in earlier comment https://github.com/cakephp/cakephp/pull/12436/files#r208102520 why allowing integer or float input for `containsNonAlphaNumeric()` doesn't make sense.

Reply:
I still think this should have been kept to allow integers to pass fine. 
Especially when those come from a different input source than just basic string POST data.